src/matcher/index.ts

```diff
@@ -71,11 +74,66 @@ export function createRouterMatcher(
     return matcherMap.get(name)
   }
 
+  function routeComponentsOptionGuard(record: RouteRecordRaw): never | void {
+    // Guard for record.components
+    // ex> components: Home -> should be like components: { default: Home, others }
+    if (!__DEV__) return // guard for development mode
+    if ('components' in record) {
+      for (const name in record.components) {
+        const rawComponent = record.components[name]
+        const { path } = record
+        // property components should be object
+        if (
+          !rawComponent ||
+          (typeof rawComponent !== 'object' &&
+            typeof rawComponent !== 'function')
+        ) {
+          warn(
+            `Components property in record  with path "${path}" is not valid. ` +
+              `Property "components" should be object. ` +
+              `ex> { components: { default: Home } }`
+          )
+          // same with navigation guards
+          throw new Error(`Invalid route record "components" property.`)
+        } else if ('then' in rawComponent) {
+          // warn if user wrote import('/component.vue') instead of () =>
+          // import('./component.vue')
+          warn(
+            `Route record "${name}" property "components" with path "${path}" is a ` +
+              `Promise instead of a function that returns a Promise. ` +
+              `Did you write "import(./MyPage.vue) instead of "` +
+              `"() => import(./MyPage.vue) ? "` +
+              `This will break in production if not fixed.`
+          )
+          // convert () => import(./Mypage.vue) to avoid
+          // warned by navigationGuards
+          record.components[name] = () => rawComponent
+        }
+        // Guard for functional components
+        if (isRouteComponent(rawComponent)) return
+        let componentPromise: Promise<
+          RouteComponent | null | undefined | void
+        > = (rawComponent as Lazy<RouteComponent>)()
+        if (!('catch' in componentPromise)) {
+          warn(
+            `Components property in record  with path "${path}" is not valid. ` +
+              `Property "components" should be object. ` +
+              `ex> { components: { default: Home } }`
+          )
+          // should throw error becuase components are not object
+          throw new Error(`Invalid route record "components" property.`)
```

Comment:
This can be simplified a lot, we only need to check if `components` is a component and tell the user they seem to have a typo and it should be `component: MyView` or `components: { default: MyView }`

Reply:
Thanks for guidance! 😄

I am confused state `simplified`. I have tried to guard cases below:
 
```javascript
{
   components: component, // 1 regular component
   components: () => import('./MyPage'), // 2 lazy loaded component
   components: { render: () => h('div', {}, 'hello') } // 3 functional component
}
```

However, I have to guard only `case 1` (case below)?

```javascript
{
   components: component, // 1
}
```

Reply:
yes